### ISSUE TITLE
47x speedup results could be incorrect since ht model is not setup well. It should be ~5x speedup. 

### ISSUE THREAD

**Issue by gyin-ai** [bytedance/lightseq#38]
1. Setup the hf model and input tensors to cuda device.
2. Warmup the model with some simple sentences

`ls_bart.py`
```
import time
import argparse

import torch
import lightseq
from transformers import BartTokenizer, BartForConditionalGeneration


def ls_bart(model, inputs):
    torch.cuda.synchronize()
    start_time = time.perf_counter()
    generated_ids = model.infer(inputs)
    torch.cuda.synchronize()
    end_time = time.perf_counter()
    return generated_ids, end_time - start_time


def hf_bart(model, inputs):
    inputs = inputs.to('cuda')
    torch.cuda.synchronize()
    start_time = time.perf_counter()
    generated_ids = model.generate(inputs, max_length=50)
    torch.cuda.synchronize()
    end_time = time.perf_counter()
    return generated_ids, end_time - start_time


def ls_generate(model, tokenizer, inputs_id):
    print("=========================lightseq=========================")
    print("lightseq generating...")
    ls_res_ids, ls_time = ls_bart(model, inputs_id)
    ls_res_ids = [ids[0] for ids in ls_res_ids[0]]
    ls_res = tokenizer.batch_decode(ls_res_ids, skip_special_tokens=True)
    print(f"lightseq time: {ls_time}s")
    print("lightseq results:")
    for sent in ls_res:
        print(sent)


def hf_generate(model, tokenizer, inputs_id):
    print("=========================huggingface=========================")
    print("huggingface generating...")
    hf_res_ids, hf_time = hf_bart(model, inputs_id)
    hf_res = tokenizer.batch_decode(hf_res_ids, skip_special_tokens=True)
    print(f"huggingface time: {hf_time}s")
    print("huggingface results:")
    for sent in hf_res:
        print(sent)


def warmup(tokenizer, ls_model, hf_model,
           sentences=["Let's do a quick warm up <mask>.", "what do you <mask> of", "that is a <mask> idea"]):
    print("======START warmup=====")
    inputs = tokenizer(sentences, return_tensors="pt", padding=True)
    inputs_id = inputs["input_ids"]

    ls_generate(ls_model, tokenizer, inputs_id)
    hf_generate(hf_model, tokenizer, inputs_id)
    print("======END warmup=====")


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument('--user_input', action="store_true")
    args = parser.parse_args()

    print("initializing bart tokenizer...")
    tokenizer = BartTokenizer.from_pretrained("facebook/bart-base")

    print("creating lightseq model...")
    ls_model = lightseq.Transformer("lightseq_bart_base.pb", 128)
    print("creating huggingface model...")
    hf_model = BartForConditionalGeneration.from_pretrained("facebook/bart-base")
    hf_model.eval()
    hf_model.to('cuda')

    warmup(tokenizer, ls_model, hf_model)

    while True:
        if args.user_input:
            sentences = [input("input the masked sentence:\n")]
        else:
            sentences = [
                "I love that girl, but <mask> does not <mask> me.",
                "She is so <mask> that I can not help glance at <mask>.",
                "Nothing's gonna <mask> my love for you.",
                "Drop everything now. Meet me in the pouring <mask>. Kiss me on the sidewalk."
            ]

        print("tokenizing the sentences...")
        inputs = tokenizer(sentences, return_tensors="pt", padding=True)
        inputs_id = inputs["input_ids"]

        ls_generate(ls_model, tokenizer, inputs_id)
        hf_generate(hf_model, tokenizer, inputs_id)

        if not args.user_input:
            break


if __name__ == "__main__":
    main()
```

output
```
tokenizing the sentences...
=========================lightseq=========================
lightseq generating...
lightseq time: 0.041104961186647415s
lightseq results:
I love that girl, but she does not love me.
She is so beautiful that I can not help glance at her.
Nothing's gonna change my love for you.
Drop everything now. Meet me in the pouring rain. Kiss me on the sidewalk.
=========================huggingface=========================
huggingface generating...
huggingface time: 0.153672456741333s
huggingface results:
I love that girl, but she does not love me.
She is so beautiful that I can not help glance at her.
Nothing's gonna change my love for you.
Drop everything now. Meet me in the pouring rain. Kiss me on the sidewalk.
```

**Comment by Taka152** [comment]
grateful for this correction, we didn’t try much about the performance of
huggingface, we’ll modify the doc soon if it’s confirmed.

gyin-ai ***@***.***>于2021年5月4日 周二14:03写道：

**Comment by gyin-ai** [comment]
@Taka152 can I ask which model lightseq uses for benchmark?

```
Which is tested on Tesla T4 with a model of Transformer-base. 
```

**Comment by godweiyang** [comment]
> 1. Setup the hf model and input tensors to cuda device.
> 2. Warmup the model with some simple sentences
> 
> `ls_bart.py`
> 
> ```
> import time
> import argparse
> 
> import torch
> import lightseq
> from transformers import BartTokenizer, BartForConditionalGeneration
> 
> 
> def ls_bart(model, inputs):
>     torch.cuda.synchronize()
>     start_time = time.perf_counter()
>     generated_ids = model.infer(inputs)
>     torch.cuda.synchronize()
>     end_time = time.perf_counter()
>     return generated_ids, end_time - start_time
> 
> 
> def hf_bart(model, inputs):
>     inputs = inputs.to('cuda')
>     torch.cuda.synchronize()
>     start_time = time.perf_counter()
>     generated_ids = model.generate(inputs, max_length=50)
>     torch.cuda.synchronize()
>     end_time = time.perf_counter()
>     return generated_ids, end_time - start_time
> 
> 
> def ls_generate(model, tokenizer, inputs_id):
>     print("=========================lightseq=========================")
>     print("lightseq generating...")
>     ls_res_ids, ls_time = ls_bart(model, inputs_id)
>     ls_res_ids = [ids[0] for ids in ls_res_ids[0]]
>     ls_res = tokenizer.batch_decode(ls_res_ids, skip_special_tokens=True)
>     print(f"lightseq time: {ls_time}s")
>     print("lightseq results:")
>     for sent in ls_res:
>         print(sent)
> 
> 
> def hf_generate(model, tokenizer, inputs_id):
>     print("=========================huggingface=========================")
>     print("huggingface generating...")
>     hf_res_ids, hf_time = hf_bart(model, inputs_id)
>     hf_res = tokenizer.batch_decode(hf_res_ids, skip_special_tokens=True)
>     print(f"huggingface time: {hf_time}s")
>     print("huggingface results:")
>     for sent in hf_res:
>         print(sent)
> 
> 
> def warmup(tokenizer, ls_model, hf_model,
>            sentences=["Let's do a quick warm up <mask>.", "what do you <mask> of", "that is a <mask> idea"]):
>     print("======START warmup=====")
>     inputs = tokenizer(sentences, return_tensors="pt", padding=True)
>     inputs_id = inputs["input_ids"]
> 
>     ls_generate(ls_model, tokenizer, inputs_id)
>     hf_generate(hf_model, tokenizer, inputs_id)
>     print("======END warmup=====")
> 
> 
> def main():
>     parser = argparse.ArgumentParser()
>     parser.add_argument('--user_input', action="store_true")
>     args = parser.parse_args()
> 
>     print("initializing bart tokenizer...")
>     tokenizer = BartTokenizer.from_pretrained("facebook/bart-base")
> 
>     print("creating lightseq model...")
>     ls_model = lightseq.Transformer("lightseq_bart_base.pb", 128)
>     print("creating huggingface model...")
>     hf_model = BartForConditionalGeneration.from_pretrained("facebook/bart-base")
>     hf_model.eval()
>     hf_model.to('cuda')
> 
>     warmup(tokenizer, ls_model, hf_model)
> 
>     while True:
>         if args.user_input:
>             sentences = [input("input the masked sentence:\n")]
>         else:
>             sentences = [
>                 "I love that girl, but <mask> does not <mask> me.",
>                 "She is so <mask> that I can not help glance at <mask>.",
>                 "Nothing's gonna <mask> my love for you.",
>                 "Drop everything now. Meet me in the pouring <mask>. Kiss me on the sidewalk."
>             ]
> 
>         print("tokenizing the sentences...")
>         inputs = tokenizer(sentences, return_tensors="pt", padding=True)
>         inputs_id = inputs["input_ids"]
> 
>         ls_generate(ls_model, tokenizer, inputs_id)
>         hf_generate(hf_model, tokenizer, inputs_id)
> 
>         if not args.user_input:
>             break
> 
> 
> if __name__ == "__main__":
>     main()
> ```
> 
> output
> 
> ```
> tokenizing the sentences...
> =========================lightseq=========================
> lightseq generating...
> lightseq time: 0.041104961186647415s
> lightseq results:
> I love that girl, but she does not love me.
> She is so beautiful that I can not help glance at her.
> Nothing's gonna change my love for you.
> Drop everything now. Meet me in the pouring rain. Kiss me on the sidewalk.
> =========================huggingface=========================
> huggingface generating...
> huggingface time: 0.153672456741333s
> huggingface results:
> I love that girl, but she does not love me.
> She is so beautiful that I can not help glance at her.
> Nothing's gonna change my love for you.
> Drop everything now. Meet me in the pouring rain. Kiss me on the sidewalk.
> ```

Thanks a lot. I have confirmed the correctness of your code. LightSeq can only get a 10x speedup.
```
=========lightseq=========
lightseq generating...
lightseq time: 0.035063838586211205s
lightseq results:
I love that girl, but she does not love me.
She is so beautiful that I can not help glance at her.
Nothing's gonna change my love for you.
Drop everything now. Meet me in the pouring rain. Kiss me on the sidewalk.
=========huggingface=========
huggingface generating...
huggingface time: 0.3570530144497752s
huggingface results:
I love that girl, but she does not love me.
She is so beautiful that I can not help glance at her.
Nothing's gonna change my love for you.
Drop everything now. Meet me in the pouring rain. Kiss me on the sidewalk.
```

**Comment by Taka152** [comment]
> @Taka152 can I ask which model lightseq uses for benchmark?
> 
> ```
> Which is tested on Tesla T4 with a model of Transformer-base. 
> ```

@gyin-ai vanilla Transformer-base in 'Attention Is All Your Need' paper, you can check the details in our arxiv paper
